### PR TITLE
R4R: add Mirror/Sync feature, implementation of BEP84

### PR DESCRIPTION
**Pull request by HaoyangLiu**
Current bind mechanism is based on the context that our community members are on the Binance Chain and they want to extend their tokens to the Binance Smart Chain. However, with the evolution of our community, things changed. In most cases, users issue bep20 on the Binance Smart Chain first without considering whether they will issue BEP2 on the Binance Chain or not,  so a new mechanism to conveniently extend BEP20 assets to the Binance Chain is required. In addition, the new mechanism will encourage users to extend their assets to the Binance Chain which is very helpful to flourish the Binance Chain community. 

Implement genesis contract for https://github.com/binance-chain/BEPs/pull/84

### CODE REVIEW

File: contracts/TokenManager.sol [hunk 1]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+    uint256 totalSupply = IBEP20(bep20Addr).totalSupply();
+    require(convertToBep2Amount(totalSupply, decimals) <= MAX_BEP2_TOTAL_SUPPLY, "too large total supply");
+    string memory name = IBEP20(bep20Addr).name();
+    bytes memory nameBytes = bytes(name);
+    require(nameBytes.length>=1 && nameBytes.length<=32, "name length must be in [1,32]");
+    string memory symbol = IBEP20(bep20Addr).symbol();
+    bytes memory symbolBytes = bytes(symbol);
+    require(symbolBytes.length>=MINIMUM_BEP20_SYMBOL_LEN && symbolBytes.length<=MAXIMUM_BEP20_SYMBOL_LEN, "symbol length must be in [3,8]");
+    for (uint8 i = 0; i < symbolBytes.length; i++) {
+      require((symbolBytes[i]>='a' && symbolBytes[i]<='a') || (symbolBytes[i]>='A' && symbolBytes[i]<='Z'), "symbol must not contain non-alphabet");
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:349]: The condition of symbol on BC is `token symbol should be alphanumeric`. So number is ok, can we allow number?

File: contracts/TokenManager.sol [hunk 2]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+      bytes32Symbol := mload(add(symbol, 32))
+    }
+    MirrorSynPackage memory mirrorSynPackage = MirrorSynPackage({
+      mirrorSender:  msg.sender,
+      bep20Addr:     bep20Addr,
+      bep20Name:     bytes32Name,
+      bep20Symbol:   bytes32Symbol,
+      bep20Supply:   totalSupply,
+      bep20Decimals: decimals,
+      mirrorFee:     mirrorFee/TEN_DECIMALS,
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:368]: use safe div please.

File: contracts/TokenManager.sol [hunk 3]
```diff
@@ -49,29 +90,71 @@ contract TokenManager is System, IApplication {
   uint8 constant public   BIND_STATUS_ALREADY_BOUND_TOKEN = 6;
   uint8 constant public   BIND_STATUS_REJECTED = 7;
 
+  uint8 constant public MIRROR_CHANNELID = 0x04;
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:93]: `MIRROR_CHANNELID` ==> `MIRROR_CHANNEL_ID `?

File: contracts/TokenManager.sol [hunk 4]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+      });
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(MIRROR_CHANNELID, encodeMirrorSynPackage(mirrorSynPackage), msg.value.sub(mirrorFee).div(TEN_DECIMALS));
+    return true;
+  }
+
+  function handleMirrorAckPackage(bytes memory msgBytes) internal {
+    (MirrorAckPackage memory mirrorAckPackage, bool decodeSuccess) = decodeMirrorAckPackage(msgBytes);
+    require(decodeSuccess, "unrecognized package");
+    mirrorPendingRecord[mirrorAckPackage.bep20Addr] = false;
+    if (mirrorAckPackage.errorCode == MIRROR_STATUS_SUCCESS ) {
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:379]: In `System.sol`, there is a code named `CODE_OK`, can we reuse that?

File: contracts/TokenManager.sol [hunk 5]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+      address(uint160(TOKEN_HUB_ADDR)).transfer(mirrorAckPackage.mirrorFee);
+      ITokenHub(TOKEN_HUB_ADDR).bindToken(mirrorAckPackage.bep2Symbol, mirrorAckPackage.bep20Addr, mirrorAckPackage.bep20Decimals);
+      boundByMirror[mirrorAckPackage.bep20Addr] = true;
+      emit mirrorSuccess(mirrorAckPackage.bep20Addr, mirrorAckPackage.bep2Symbol);
+      return;
+    } else {
+      emit mirrorFailure(mirrorAckPackage.bep20Addr, mirrorAckPackage.errorCode);
+    }
+    (bool success, ) = mirrorAckPackage.mirrorSender.call{gas: MAX_GAS_FOR_TRANSFER_BNB, value: mirrorAckPackage.mirrorFee}("");
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:388]: Can you put the transfer BNB to user account input the logic of else? Make the logic easy to readn

Comment (guagualvcha) [inline contracts/TokenManager.sol:388]: Miss safe mul `TEN_DECIMALS` here for the `mirrorFee`?

File: contracts/TokenManager.sol [hunk 6]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+    if (!success) {
+      address(uint160(SYSTEM_REWARD_ADDR)).transfer(mirrorAckPackage.mirrorFee);
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:390]: So do here

File: contracts/TokenManager.sol [hunk 7]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+    }
+  }
+
+  function handleMirrorFailAckPackage(bytes memory msgBytes) internal {
+    (MirrorSynPackage memory mirrorSynPackage, bool decodeSuccess) = decodeMirrorSynPackage(msgBytes);
+    require(decodeSuccess, "unrecognized package");
+    mirrorPendingRecord[mirrorSynPackage.bep20Addr] = false;
+    (bool success, ) = mirrorSynPackage.mirrorSender.call{gas: MAX_GAS_FOR_TRANSFER_BNB, value: mirrorSynPackage.mirrorFee*TEN_DECIMALS}("");
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:398]: use safe mul please.

File: contracts/TokenManager.sol [hunk 8]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+      });
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(SYNC_CHANNELID, encodeSyncSynPackage(syncSynPackage), msg.value.sub(syncFee).div(TEN_DECIMALS));
+    return true;
+  }
+
+  function handleSyncAckPackage(bytes memory msgBytes) internal {
+    (SyncAckPackage memory syncAckPackage, bool decodeSuccess) = decodeSyncAckPackage(msgBytes);
+    require(decodeSuccess, "unrecognized package");
+    if (syncAckPackage.errorCode == SYNC_STATUS_SUCCESS ) {
+      address(uint160(TOKEN_HUB_ADDR)).transfer(syncAckPackage.syncFee);
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:483]: Why transfer syncFee again, seems already transfer when start `sync`?

File: contracts/TokenManager.sol [hunk 9]
```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
...
+    require(bep2Symbol != bytes32(0x00), "not bound");
+    require(boundByMirror[bep20Addr], "not bound by mirror");
+    uint256 miniRelayFee = ITokenHub(TOKEN_HUB_ADDR).getMiniRelayFee();
+    require(msg.value%TEN_DECIMALS == 0 && msg.value>=syncFee.add(miniRelayFee), "msg.value must be N * 1e10 and no less sum of miniRelayFee and syncFee");
+    require(expireTime>=block.timestamp + 120 && expireTime <= block.timestamp + 86400, "expireTime must be two minutes later and one day earlier");
+    uint256 totalSupply = IBEP20(bep20Addr).totalSupply();
+    uint8 decimals = IBEP20(bep20Addr).decimals();
+    require(convertToBep2Amount(totalSupply, decimals) <= MAX_BEP2_TOTAL_SUPPLY, "too large total supply");
+
+    address(uint160(TOKEN_HUB_ADDR)).transfer(msg.value.sub(syncFee));
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:466]: Shoud transfer relayer fee here, not `syncFee `.

File: contracts/TokenManager.sol [hunk 10]
```diff
@@ -260,4 +600,11 @@ contract TokenManager is System, IApplication {
     }
     return symbolMatch;
   }
+
+  function convertToBep2Amount(uint256 amount, uint256 bep20TokenDecimals) internal pure returns (uint256) {
+    if (bep20TokenDecimals > BEP2_TOKEN_DECIMALS) {
+      return amount.div(10**(bep20TokenDecimals-BEP2_TOKEN_DECIMALS));
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:606]: What will happen if `bep20TokenDecimals-BEP2_TOKEN_DECIMALS` is quite large that overflow? Can you have a test case for this?

File: contracts/TokenManager.sol [hunk 11]
```diff
@@ -120,6 +203,7 @@ contract TokenManager is System, IApplication {
   }
 
   function approveBind(address contractAddr, string memory bep2Symbol) payable public returns (bool) {
+    require(!mirrorPendingRecord[contractAddr], "the bep20 token is in mirror pending status");
     bytes32 bep2TokenSymbol = bep2TokenSymbolConvert(bep2Symbol);
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:207]: Please handle if mirror succeed.

File: contracts/TokenManager.sol (hunk 9 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:466]: What relayer fee? The minimum relayer fee? The fee should not be refunded.

File: contracts/TokenManager.sol (hunk 7 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:398]: The mirrorFee is from the mirror. It won't be overflow.

File: contracts/TokenManager.sol (hunk 5 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:388]: I need to reduce the code size.

File: contracts/TokenManager.sol (hunk 2 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:368]: safe div just ensure the `TEN_DECIMALS ` is not zero. Not necessary here

File: contracts/TokenManager.sol (hunk 1 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:349]: BC require all symbol is alphanumeric only

File: contracts/TokenManager.sol (hunk 5 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:388]: BC will convert the mirrorFee to 18 decimals. So no need to mul `TEN_DECIMALS `

File: contracts/TokenManager.sol (hunk 6 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:390]: BC will convert the mirrorFee to 18 decimals. So no need to mul `TEN_DECIMALS`

File: contracts/TokenManager.sol (hunk 10 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:606]: Good idea.

File: contracts/TokenManager.sol (hunk 11 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:207]: Good idea

File: contracts/TokenManager.sol (hunk 4 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:379]: Good idea.

File: contracts/TokenManager.sol (hunk 1 above)
Comment (guagualvcha) [reply contracts/TokenManager.sol:349]: But the implementation here: `(symbolBytes[i]>='a' && symbolBytes[i]<='a')` is confusing and number is invalid.

File: contracts/TokenManager.sol (hunk 2 above)
Comment (guagualvcha) [reply contracts/TokenManager.sol:368]: Plesae make the uniform style.  Either change line 371 or here.

File: contracts/TokenManager.sol (hunk 5 above)
Comment (guagualvcha) [reply contracts/TokenManager.sol:388]: Put the code outside `if`  will reduce code size?

File: contracts/TokenManager.sol (hunk 9 above)
Comment (guagualvcha) [reply contracts/TokenManager.sol:466]: I mean here did not transfer the `minimum relayer fee`.

File: contracts/TokenManager.sol (hunk 11 above)
Comment (guagualvcha) [reply contracts/TokenManager.sol:207]: If user already mirror success, looks like apporveBind will success so far, won't it confict?

File: contracts/TokenManager.sol (hunk 1 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:347]: My mistake, it should be `(symbolBytes[i]>='a' && symbolBytes[i]<='z')`

File: contracts/TokenManager.sol (hunk 11 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol:205]: `apporveBind ` will return error code for already bound

File: contracts/TokenManager.sol (hunk 5 above)
Comment (HaoyangLiu) [reply contracts/TokenManager.sol]: Previous I added many `else if` to enumerate all error reason. Later I removed that to recude code size. Just forgot to add it under `if`

File: contracts/TokenManager.sol [hunk 12]
```diff
@@ -189,6 +271,262 @@ contract TokenManager is System, IApplication {
...
+    } else {
+      (bool success, ) = mirrorAckPackage.mirrorSender.call{gas: MAX_GAS_FOR_TRANSFER_BNB, value: mirrorAckPackage.mirrorFee}("");
+      if (!success) {
+        address(uint160(SYSTEM_REWARD_ADDR)).transfer(mirrorAckPackage.mirrorFee);
+      }
+      emit mirrorFailure(mirrorAckPackage.bep20Addr, mirrorAckPackage.errorCode);
+    }
+  }
+
+  function handleMirrorFailAckPackage(bytes memory msgBytes) internal {
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:392]: Can we emit `mirrorFailure` here.

File: contracts/TokenManager.sol [hunk 13]
```diff
@@ -189,6 +271,262 @@ contract TokenManager is System, IApplication {
...
+      emit syncFailure(syncAckPackage.bep20Addr, syncAckPackage.errorCode);
+    }
+    (bool success, ) = syncAckPackage.syncSender.call{gas: MAX_GAS_FOR_TRANSFER_BNB, value: syncAckPackage.syncFee}("");
+    if (!success) {
+      address(uint160(SYSTEM_REWARD_ADDR)).transfer(syncAckPackage.syncFee);
+    }
+  }
+
+  function handleSyncFailAckPackage(bytes memory msgBytes) internal {
+    (SyncSynPackage memory syncSynPackage, bool decodeSuccess) = decodeSyncSynPackage(msgBytes);
```
Comment (guagualvcha) [inline contracts/TokenManager.sol:494]: emit `syncFailure` here please.

**Review by guagualvcha — APPROVED** [review]

**Review by yutianwu — APPROVED** [review]

**Pull request merged**